metals/src/main/scala/scala/meta/internal/metals/Diagnostics.scala

```diff
@@ -312,4 +313,30 @@ final class Diagnostics(
     toPublish
   }
 
+  private def adjustedDiagnosticData(
+      diagnostic: l.Diagnostic,
+      edit: TokenEditDistance,
+  ): Option[Object] =
+    diagnostic match {
+      case ScalacDiagnostic.ScalaDiagnostic(Left(textEdit)) =>
+        edit
+          .toRevised(
+            textEdit,
+            shouldAdjustWithinToken(diagnostic),
+            fallbackToNearest = false,
+          )
+          .map(_.toJsonObject)
+      case ScalacDiagnostic.ScalaDiagnostic(Right(scalaDiagnostic)) =>
+        edit
+          .toRevised(
+            scalaDiagnostic,
+            shouldAdjustWithinToken(diagnostic),
+            fallbackToNearest = false,
+          )
+          .map(_.toJsonObject)
+      case _ => Some(diagnostic.getData())
+    }
+
+  private def shouldAdjustWithinToken(diagnostic: l.Diagnostic): Boolean =
+    diagnostic.getSource() == "scala-cli"
```

Comment:
Minor point, but just looking over this again, I'm not sure this is very future-proof - it could result in misalignment of diagnostics if `scala-cli` were to implement more diagnostics which weren't within comments.  Perhaps we could match on [code](https://build-server-protocol.github.io/docs/specification/#diagnostic) to determine this (though I don't think this is currently set by scala-cli?)?

Reply:
> it could result in misalignment of diagnostics

With the current implementation we match on any tokens, so we don't single out comments. In which case would we have misaligned diagnostics?

However, coming to think about this line. we probably don't need ` diagnostic.getSource()  == "scala-cli"` check at all. I think we would want to adjust edits coming from any build server really. You could remove it.

Reply:
> With the current implementation we match on any tokens, so we don't single out comments. In which case would we have misaligned diagnostics?

Sorry, I meant in the case if `scala-cli` were to issue some another custom diagnostic with source set to `scala-cli` (so no case atm afaik), but it's a minor point :slightly_smiling_face:  

> However, coming to think about this line. we probably don't need  diagnostic.getSource()  == "scala-cli" check at all. I think we would want to adjust edits coming from any build server really. You could remove it.

Yeah I originally had it apply to everything, but it gave poor results in some circumstances, for example:
![not-used](https://github.com/scalameta/metals/assets/17688577/a0f2356e-752c-4fb1-96e8-4cf7f206e32a)

Reply:
I am thinking of doing a follow up removing the check, but if you could clarify why that broke for you that would be awesome

Reply:
Ach wait no, that a totally different thing than I though, ignore me